src/expr/src/typecheck.rs

```diff
@@ -0,0 +1,777 @@
+                    }
+                }
+
+                body.typecheck(&mut ctx)
+            }
+            ArrangeBy { input, keys } => {
+                let t_in = input.typecheck(ctx)?;
+
+                for cols in keys {
+                    for col in cols {
+                        let _ = col.typecheck(self, &t_in)?;
+                    }
+                }
+
+                Ok(t_in)
+            }
+        }
+    }
+
+    /// Traverses a term to collect the types of given ids.
+    ///
+    /// LetRec doesn't have type info stored in it. Until we change the MIR to track that information explicitly, we have to rebuild it from looking at the term.
+    fn collect_recursive_variable_types(
+        &self,
+        ids: &[LocalId],
+        ctx: &mut Ctx,
+    ) -> Result<(), TypeError> {
+        match self {
+            MirRelationExpr::Get {
+                id: Id::Local(id),
+                typ,
+            } => {
+                if !ids.contains(id) {
+                    return Ok(());
+                }
+
+                let id = Id::Local(id.clone());
+                if let Some(ctx_typ) = ctx.get_mut(&id) {
+                    for (base_col, input_col) in ctx_typ.iter_mut().zip_eq(typ.column_types.iter())
+                    {
+                        *base_col =
+                            base_col
+                                .union(&input_col)
+                                .map_err(|e| TypeError::MismatchColumn {
+                                    source: self,
+                                    got: input_col.clone(),
+                                    expected: base_col.clone(),
+                                    message: format!(
+                                        "couldn't compute union of collected column types: {:?}",
+                                        e
+                                    ),
+                                })?;
+                    }
+                } else {
+                    ctx.insert(id, typ.column_types.clone());
+                }
+            }
+            MirRelationExpr::Get {
+                id: Id::Global(..), ..
+            }
+            | MirRelationExpr::Constant { .. } => (),
+            MirRelationExpr::Let { id, value, body } => {
+                value.collect_recursive_variable_types(ids, ctx)?;
+
+                // we've shadowed the id
+                if ids.contains(id) {
+                    warn!("id {} shadowed in {:#?}", id, self);
+                    return Ok(());
+                }
+
+                body.collect_recursive_variable_types(ids, ctx)?;
+            }
+            MirRelationExpr::LetRec {
+                ids: inner_ids,
+                values,
+                body,
+            } => {
+                for inner_id in inner_ids {
+                    if ids.contains(inner_id) {
+                        panic!("let recs shadowing other let recs (inner id {} appears in outer ids {:?}", inner_id, ids);
```

Comment:
I'd recommend using an explicit error here, or add a comment why this is an invariant violation that we cannot recover from. Panicking takes down `environmentd`, which is not great.

Reply:
In that case I'm tempted to treat this as a `warn!`, like other shadowing violations.

Reply:
Ideally, we would want such errors to be as explicit as possible so that our fuzzing tools can run into them without requiring additional instrumentation. There are two testing-friendly options:

- throw an "internal error" back to the SQL client
- use a soft assertion to only panic in the CI and during testing, but never in production. Note that our CI and fuzzers run against Release builds

@def- FYI.

Reply:
Changing this to be an explicit `TypeError`, which will turn into `TransformError::Internal`.

Right now we just report the shadowing ID and not the shadowed ID... it'd take some re-engineering to track both. We dump the whole outer term and the shadowing term in the message, which should be enough to localize things.

Unrelated: it would be nice to unify all of the `&[ColumnType]` methods in one place, with a nice type differencer that gives concise error messages (e.g., "these types differ in nullability of column 17" rather than giving the user a "find the difference" puzzle).